bittensor/_neuron/text/advanced_server/nucleus_impl.py

```diff
@@ -87,7 +87,56 @@ def __init__(self,
         
         # -- keeps track of gradients applied
         self.backward_gradients = 0 
-        
+        self.set_fine_tuning_params()
+
+    def set_fine_tuning_params(self):
+        r''' Set to tune only the parameter of the last layer
+        '''
+        def find_last_layer(model):    
+            r''' Recursively find the last layer in a nn.ModuleList
+            '''
+            reverted_child_list = [(name, child) for name, child in model.named_children()]
+            reverted_child_list.reverse()
+
+            for name, child in reverted_child_list:    
+                if isinstance(child, nn.ModuleList):
+                    if self.config.neuron.finetune.num_layers > len(child):
+                        logger.warning(f'Number of finetune layers was set higher then the layers avaliable {len(child)}')
+                        return None
+                    return (name + '.' +str(len(child) - self.config.neuron.finetune.num_layers))
+                
+            for name, child in reverted_child_list:    
+                name_ = find_last_layer(child)
+                if name_ != None:
+                    return (name+'.'+ name_)
+
+            return (None)     
+
+        if self.config.neuron.finetune.layer_name == None:
+            last_layer_name = find_last_layer(self.pre_model)
+        else:
+            last_layer_name = self.config.neuron.finetune.layer_name
+
+        reached_last_layer = False
+
+        # set the non-last layer parameters not to require grads
+        if (not self.config.neuron.finetune.all) and (last_layer_name != None):
+            logger.success(f'Set to finetune layer {last_layer_name} and onwards')
+            for name, param in self.pre_model.named_parameters():
+                if last_layer_name in name or reached_last_layer == True:
+                    param.requires_grad = True
+                    reached_last_layer = True
+                else:
+                    param.requires_grad = False
```

Comment:
I find this section a bit confusing to follow, and I expect it will be challenging to refactor or debug later. Could we try to simplify the logic a bit?